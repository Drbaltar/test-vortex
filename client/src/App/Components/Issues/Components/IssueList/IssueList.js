import React from 'react';
import IssueEntry from './Components/IssueEntry/IssueEntry';

const IssueList = (props) => {
    const { list, selectEntry } = props;

    return(
        <div>
            <p id='issue-list-message'>{getErrorMessage(list.length)}</p>
            {getIssueList(list, selectEntry)}
        </div>
    );
};

const getErrorMessage = (length) => {
    if (length === 0) {
        return 'No Issues Found!';
    } else {
        return `${length} Open Issue${length > 1 ? 's' : ''} Found!`;
    }
};

const getIssueList = (list, selectEntry) => {
    return list.map((entry, index) =>
        <IssueEntry issue={entry} key={index} select={selectEntry} index={index}/>
    );
};

IssueList.defaultProps = {
    list: []
};

export default IssueList;
