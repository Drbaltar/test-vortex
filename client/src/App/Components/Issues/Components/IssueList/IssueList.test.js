import React from 'react';
import { shallow } from 'enzyme';
import IssueList from './IssueList';

const singleEntry = [
    {
        issue_type: 'Application Issue',
        issue_category: 'Application Crash',
        issue_description: 'The application keeps crashing!',
        submitted_by: 'kmccain'
    }
];

const exampleList = [
    {
        issue_type: 'Application Issue',
        issue_category: 'Application Crash',
        issue_description: 'The application keeps crashing!',
        submitted_by: 'kmccain'
    },
    {
        issue_type: 'Application Issue',
        issue_category: 'Entry/List Not Populating Information',
        issue_description: 'The question list is not populating!',
        submitted_by: 'kmccain'
    }
];

const mockSelectFunction = jest.fn();

describe('IssueList', () => {
    const wrapper = shallow(<IssueList />);

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    describe('when an empty list is passed in', () => {
        beforeAll(() => {
            wrapper.setProps({ list: [] });
        });

        it('returns a message that there are no issues found if passed an empty list', () => {
            expect(wrapper.find('#issue-list-message').text()).toBe('No Issues Found!');
        });
    });

    describe('when a single item list is passed in', () => {
        beforeAll(() => {
            wrapper.setProps({
                list: singleEntry,
                selectEntry: mockSelectFunction
            });
        });

        it('returns a message that there is one issue found', () => {
            expect(wrapper.find('#issue-list-message').text()).toBe('1 Open Issue Found!');
        });
    });

    describe('when a non-empty list is passed in', () => {
        beforeAll(() => {
            wrapper.setProps({
                list: exampleList,
                selectEntry: mockSelectFunction
            });
        });

        it('returns a message that there are `n` number of issues found', () => {
            expect(wrapper.find('#issue-list-message').text()).toBe(`${exampleList.length} Open Issues Found!`);
        });

        it('creates the correct number of list items', () => {
            expect(wrapper.find('IssueEntry').length).toEqual(exampleList.length);
        });

        it('passes the issue entry to the IssueEntry component', () => {
            expect(wrapper.find('IssueEntry').first().prop('issue')).toEqual(exampleList[0]);
        });

        it('passes the index entry to the IssueEntry component', () => {
            expect(wrapper.find('IssueEntry').at(0).prop('index')).toEqual(0);
            expect(wrapper.find('IssueEntry').at(1).prop('index')).toEqual(1);
        });

        it('passes the function for selecting an entry to the IssueEntry component', () => {
            expect(wrapper.find('IssueEntry').first().prop('select')).toEqual(mockSelectFunction);
        });
    });

});