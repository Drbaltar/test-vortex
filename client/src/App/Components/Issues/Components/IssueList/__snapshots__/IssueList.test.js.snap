// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueList renders correctly 1`] = `
<div>
  <p
    id="issue-list-message"
  >
    No Issues Found!
  </p>
</div>
`;
